src/app.rs

```diff
@@ -227,75 +239,129 @@ impl Application {
+                                                api_server.update_config(topology.config());
+                                            }
+
+                                            emit!(VectorReloaded { config_paths: &config_paths })
+                                        },
+                                        Ok(false) => emit!(VectorReloadFailed),
+                                        // Trigger graceful shutdown for what remains of the topology
+                                        Err(()) => {
+                                            emit!(VectorReloadFailed);
+                                            emit!(VectorRecoveryFailed);
+                                            break Control::Shutdown;
                                         }
+                                    }
+                                    sources_finished = topology.sources_finished();
+                            }
+                            // Reload a configuration from the filesystem. If a new provider is
+                            // given, the previous drops out of scope and will be cleaned up.
+                            Control::Reload => {
+                                // Reload paths
+                                config_paths = config::process_paths(&opts.config_paths_with_formats()).unwrap_or(config_paths);
+                                // Reload config
+                                let new_config = config::load_from_paths(&config_paths).map_err(handle_config_errors).ok();
+
+                                if let Some(mut new_config) = new_config {
+                                    #[cfg(feature = "providers")]
+                                    // If there's a new provider, (re)instantiate it.
+                                    if new_config.provider.is_some() {
+                                        _provider = config::provider::init_provider(new_config.provider.take())
+                                            .await
+                                            .map(|provider| controller.with_shutdown(ReceiverStream::new(provider)));
+                                    }
 
-                                        emit!(VectorReloaded { config_paths: &config_paths })
-                                    },
-                                    Ok(false) => emit!(VectorReloadFailed),
-                                    // Trigger graceful shutdown for what remains of the topology
-                                    Err(()) => {
-                                        emit!(VectorReloadFailed);
-                                        emit!(VectorRecoveryFailed);
-                                        break SignalTo::Shutdown;
+                                    new_config.healthchecks.set_require_healthy(opts.require_healthy);
+                                    match topology
+                                        .reload_config_and_respawn(new_config)
```

Comment:
> 2\. Providers being both the source of configs and a field on them seems like the wrong model. I'd suggest breaking apart the ideas of a top-level config (i.e. provider, global options, etc) and the topology config (i.e. sources, transforms, sinks). We could leave them as siblings on `ConfigBuilder` if it's simpler for deserialization, but we don't need to extend that model to `Config` itself.

I haven't had a chance to look at this PR yet, but this particular point prompted the thought that we could leave it out of the config altogether and make it a CLI flag like `--remote-config http://example.com` if having it in the config is causing issues. I can see this might simplify the user's mental model that it is one or the other rather than raising any questions about what happens when you have both a `provider` and components in the file.

Reply:
> I haven't had a chance to look at this PR yet, but this particular point prompted the thought that we could leave it out of the config altogether and make it a CLI flag like --remote-config http://example.com if having it in the config is causing issues

I think that would be a convenient shortcut for the `http` provider. Future providers are likely to have more elaborate flags. The Datadog pipelines provide, for example, lacks a URL and instead has an API key, pipeline ID, and some other pieces.

There's also a need to still set other top-level options, such as `[api]`, `data_dir`, etc. Remote loading is only concerned with topology config; the remaining keys are all still present in config. We could also move those to CLI flags too (where there's not already a counterpart), but I think it'd still be good to support specifying a provider in a config file for parity.

> This section seems particularly confusing. We reload a config from disk, optionally spawn a new provider, and then reload the topology, which might not include any sources/sinks/transforms if the config was a provider? And then presumably the newly spawned provider will send a config via the controller to be reloaded in a subsequent loop?

The current steps are...

1. We load a config from disk. This can either contain sources/transforms/sinks, _or_ a provider (if both-- validation balks / Vector exits)
2. In the case of a "full" config -- this is already applied against topology, so nothing else happens. Signal processing continues as normal for later disk reloads.
3. In the case of a "provider" config -- the provider is spawned. When it's ready, it sends a `Control::Config(config)` containing the string serialised to a `Config`. This is done in the provider implementation, since what "serialising" means could be different to each provider.
4. If a new config is given (from the provider), its validated and loaded. A remote config cannot drive top-level keys such as api or data_dir, so just sources/transforms/sinks are augmented and reloaded.

> I think this is a strong indication that (1) we're trying to do too much, and (2) we're not adequately separating concerns here. We've modeled providers as just another field on configs, alongside sources/sinks/transforms, which is leading to awkwardness like this where clearly one of these paths is not doing anything.

This was modelled to allow for a provider to "stand in" for sources/transforms/sinks, which are then loaded some time later.

I'd like to improve the parts you feel are awkward. Do you have any suggestions for how you'd like to see the above being adapted?

> I don't think we need to handle dynamically reloading providers or switching between file config and provider configs. Those are cases you should just restart Vector imo. This should help us avoid the weird nested reloading, where we maintain all the complexity of the previous file-based reloading and just layer providers on top.

I did it this way to try not to break current behaviour. If using a `[provider]` means the file watcher no longer works, that felt like we'd be violating a user expectation.

A `Control::Reload` reloads the _entire_ config from disk... and a `Control::Config(config)` supplies a new one (typically, from a provider.)

> Providers being both the source of configs and a field on them seems like the wrong model

It wasn't immediately obvious how to work around this condition. Providers are defined _inside_ a config, and represent a stand-in for components loaded some time later, so by definition, they're both config and contributors to it.

I can try model topology as its own parent key to avoid everything (confusingly) existing at the top-level as current.

Reply:
> Future providers are likely to have more elaborate flags. The Datadog pipelines provide, for example, lacks a URL and instead has an API key, pipeline ID, and some other pieces

I think this could be encoded as a URL too:

```
--remote-config datadog://somehost.datadoghq.com/<PIPELINE_ID>?key=<KEY>&refresh_interval=60s
```

Just a thought anyway. We might be able to get away with a single URL flag.

>  Remote loading is only concerned with topology config; the remaining keys are all still present in config

As a naive user, I find this a little surprising that this config isn't also loaded by the remote config provider. I would expect it to load everything, including global config (`data_dir`, `api`, etc.). I think this will be useful for orchestration, but moreover, simplifies the mental model of how this works.

Reply:
The URL coding is an interesting idea. I could see that working with certain providers.

For Datadog specifically, I don't think we're going to expose the endpoint as a configurable URL, so I'm not sure there's a universal format here that works with every provider.

> As a naive user, I find this a little surprising that this config isn't also loaded by the remote config provider. I would expect it to load everything, including global config (data_dir, api, etc.). I think this will be useful for orchestration, but moreover, simplifies the mental model of how this works.

There was some internal discussion about this (no link to the thread, unfortunately) but the general thinking was that these are "host"-level config and out of scope for topology.

Setting whether the API is enabled or not, for example, isn't really something a Datadog pipeline config provider would need to know or design for. Its immediate scope is just pipeline processing. There's also a likelihood we'd want the API enabled _before_ returning config, so users could query health and determine whether remote loading succeeded.

For those kinds of scenarios, it seemed easier to scope remote loading to topology -- at least initially.

I could see an argument for allowing those top-level keys to be merged. So `api.enabled = false` initially could become `api.enabled = true` if it's initially omitted but the remote config contains it.

Reply:
👍  Thanks for the discussion!

> For Datadog specifically, I don't think we're going to expose the endpoint as a configurable URL, so I'm not sure there's a universal format here that works with every provider.

If we don't want to expose a hostname it could just be `datadog://<pipeline_id>` etc. I think the URL scheme is flexible enough to accommodate different patterns. I think it is likely that we may want to allow configuration of the host there though, as we do with `dd-agent`:

https://github.com/DataDog/datadog-agent/blob/cb0e9c15ba8629a507a8225d332590f01da8779f/pkg/config/config_template.yaml#L16-L24

> > As a naive user, I find this a little surprising that this config isn't also loaded by the remote config provider. I would expect it to load everything, including global config (data_dir, api, etc.). I think this will be useful for orchestration, but moreover, simplifies the mental model of how this works.
> 
> There was some internal discussion about this (no link to the thread, unfortunately) but the general thinking was that these are "host"-level config and out of scope for topology.
>
> Setting whether the API is enabled or not, for example, isn't really something a Datadog pipeline config provider would need to know or design for. Its immediate scope is just pipeline processing. There's also a likelihood we'd want the API enabled before returning config, so users could query health and determine whether remote loading succeeded.

True, I think we could go this route, but we'll need to be very clear about it and maybe use different terminology (like `topology_provider` rather than `provider`). My personal preference would be for it the provider to be able to provide _all_ of the config. I think this is a simpler mental model, even if the Datadog provider won't take advantage of all of it.

Reply:
> If we don't want to expose a hostname it could just be datadog://<pipeline_id> etc. I think the URL scheme is flexible enough to accommodate different patterns.

I could see an argument for keeping provider completely outside of config. I can also see the point of using a URL for the purposes of scoping remote config to a single field, and preventing flag sprawl.

But I think it has a fair number of downsides too:

1. It makes for potentially large/cumbersome URIs. For example, if the `http` provider allows for retry/backup a la `TowerRequestConfig](https://github.com/timberio/vector/blob/2b150a3e4feea493dde4cd85925458812208fd73/src/sinks/util/service.rs#L167-L184), a typical URL might look like this:

```
--remote-config "https://example.com/path/to/url?in_flight_limit=x&timeout_secs=x& rate_limit_duration_secs=x&rate_limit_num=x&retry_attempts=x&retry_max_duration_secs=x&retry_initial_backoff_secs=x&request[headers]=\"Authorization: Bearer <Some API Key Here>\""
```

2. It's not clear where a query string ends and internal settings begin.

3. If `<protocol>://` is the format, `http` and `https` might be separate providers by default. Likewise, if two providers use a HTTP URL, how do we know which provider it relates to?

4. It encourages passing API keys as command line arguments, which may make it easier for credentials to be exposed in OS logs.

5. It's outside of our usual TOML/YAML/JSON config, despite having a fair amount of overlap with things like the http source/sink. Introducing a protocol-specific connection string seems less clear to me than simply having a `[provider]` key.

6. Since it removes the need for a config altogether (which I admit, could be a shortcut/benefit), it doesn't allow for setting API or global options at all. As mentioned, certain providers won't have any concept or concern with those settings, or even any notion of what Vector config looks like. The Datadog provider, for example, won't return the usual TOML-- it'll be Datadog-specific pipeline JSON that the Vector provider will be responsible for translating to config. In that scenario, there's no concept or concern for things like `data_dir`, `api.enabled`, etc.

Reply:
> I think we could go this route, but we'll need to be very clear about it and maybe use different terminology (like topology_provider rather than provider). My personal preference would be for it the provider to be able to provide all of the config.

I'm not opposed to a provider overwriting all config, but I think the important distinction here is _scope_.

A `http` provider, for example, is a 1:1 replacement for local TOML/JSON/YAML. It probably makes more sense for a HTTP endpoint to be able to specify an `[api]` block or `data_dir` key to and effectively allow for 'empty' config on the connecting host.

A Datadog pipelines provider, OTOH, has no concept of Vector or Vector config. It's not concerned with where data or file checkpoints are stored, or whether or not the API is enabled. It doesn't return TOML: it responds with a payload that looks like the [Log Pipelines API](https://docs.datadoghq.com/api/latest/logs-pipelines/), with extended meta. It's the job of the provider to transform that config internally, into a representation that reflects a DD pipeline.

In that scenario, `[api]` or `data_dir` has no representation. Those are host/agent-level config settings.

That's why the [current RFC](https://github.com/timberio/vector-enterprise/pull/52) (note: non-public) favours allowing top-level config to be set alongside a `[provider]`. It's not always going to be relevant for the provider to know/care about settings that are specific to a running Vector instance.

I think the one difference that enabling those keys to be set by a provider (where relevant-- like `http`) would simply be allowing them to be overridden.

Locally, I'm working on refactoring that enables exactly that, as well as a synchronous (first) load of the config per other comments.